hera_cal/utils.py

```diff
@@ -731,3 +731,114 @@ def synthesize_ant_flags(flags, threshold=0.0):
+            is provided, this is the sum of the input chisq and the calculated chisq from all unflagged
+            data-to-model comparisons possible given their overlapping baselines. If by_antpol is True,
+            instead returns a dictionary that maps antenna polarization strings to these numpy arrays.
+        nObs: numpy array with the integer number of unflagged data-to-model comparisons that go into
+            each time and frequency of the chisq calculation. If nObs is specified, this updates that
+            with a count of any new unflagged data-to-model comparisons. If by_antpol is True, instead
+            returns a dictionary that maps antenna polarization strings to these numpy arrays.
+        chisq_per_ant: dictionary mapping ant-pol keys like (1,'x') to chisq per antenna, computed as
+            above but keeping i fixed and varying only j. If chisq_per_ant is specified, this adds in
+            new chisq calculations that include this antenna
+        nObs_per_ant: dictionary mapping ant-pol keys like (1,'x') to the integer number of unflagged
+            data-to-model comparisons that go into each time and frequency of the per-antenna chisq
+            calculation. If nObs_per_ant, this is updated to include all new unflagged observations.
+    """
+    # build containers for chisq and nObs if not supplied
+    if chisq is None and nObs is None:
+        if by_antpol:
+            chisq = {}
+            nObs = {}
+        else:
+            chisq = np.zeros(list(data.values())[0].shape, dtype=float)
+            nObs = np.zeros(list(data.values())[0].shape, dtype=int)
+    elif (chisq is None) ^ (nObs is None):
+        raise ValueError('Both chisq and nObs must be specified or nor neither can be.')
+
+    # build containers for chisq_per_ant and nObs_per_ant if not supplied
+    if chisq_per_ant is None and nObs_per_ant is None:
+        chisq_per_ant = {}
+        nObs_per_ant = {}
+    elif (chisq_per_ant is None) ^ (nObs_per_ant is None):
+        raise ValueError('Both chisq_per_ant and nObs_per_ant must be specified or nor neither can be.')
+
+    for bl in data.keys():
+        if model.has_key(bl) and data_wgts.has_key(bl) and (not by_antpol or bl[2][0] == bl[2][1]):
+            ant1, ant2 = (bl[0], bl[2][0]), (bl[1], bl[2][1])
+
+            # multiply model by gains if they are supplied
+            if gains is not None:
+                model_here = model[bl] * gains[ant1] * np.conj(gains[ant2])
+            else:
+                model_here = copy.deepcopy(model[bl])
+
+            # include gain flags in data weights
+            assert np.isrealobj(data_wgts[bl])
+            if gain_flags is not None:
+                wgts = data_wgts[bl] * ~(gain_flags[ant1]) * ~(gain_flags[ant2])
+            else:
+                wgts = copy.deepcopy(data_wgts[bl])
+
+            # calculate chi^2
+            chisq_here = np.array(np.abs(model_here - data[bl]) * wgts, dtype=float)
```

Comment:
perhaps use `np.asarray` here? the `np.array` means that the data will be copied, whereas `np.asarray` will not make a copy if the data already conforms to the desired standard.

Reply:
and perhaps use `dtype=np.float64` just to be precise as to the data type

Reply:
Done.